---
title: "Wolbachia screening"
author: "Hans Schrieke and Julie Reveillaud"
date: "`r Sys.Date()`"
output: 
  rmdformats::readthedown:
    lightbox: true
    gallery: false
    fig_caption: true
    highlight: tango
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Paths and libraries setting

```{r, warning=FALSE, message=FALSE}
# load main packages, paths and custom functions
source("../../source/main_packages.R")
source("../../source/functions.R")

# load supplementary packages
packages <- c("Biostrings", "seqinr", "kableExtra")
invisible(lapply(packages, require, character.only = TRUE))

# create output folders if needed
if (!dir.exists("../../../output/2_Oligotyping/2E")) {dir.create("../../../output/2_Oligotyping/2E")}
```

# Import fasta from oligotyping results

```{r warning=FALSE}
fasta_oligo <- readDNAStringSet("../../../output/2_Oligotyping/2A/Wolbachia/2A_oligotyping_Wolbachia_sequences-c2-s1-a0.0-A0-M10/OLIGO-REPRESENTATIVES.fasta")
fasta_oligo <- read.fasta("../../../output/2_Oligotyping/2A/Wolbachia/2A_oligotyping_Wolbachia_sequences-c2-s1-a0.0-A0-M10/OLIGO-REPRESENTATIVES.fasta")
```

# Extract fragment from first Wolbachia oligotype (test)

```{r}
myfrag <- getFrag(fasta_oligo[[1]], begin = 250, end = 279)
```

# Extract fragment from each Wolbachia oligotype

```{r}

# loop to extract each fragment from Wolbachia oligotypes
names <- c()
fragments <- c()
for(i in 1:length(fasta_oligo)){
  
  # AT, GT, AG, TT, GG, AC
  if(i!=5){
    names <- c(names, names(fasta_oligo)[i])
    myfrag <- getFrag(fasta_oligo[[i]], begin = 250, end = 279)
    myseq <- getSequence(myfrag) %>% paste(collapse = "") %>% toupper()
    fragments <- c(fragments, myseq)
  }
  
  # CC because deletion
  if(i==5){
    names <- c(names, names(fasta_oligo)[i])
    myfrag <- getFrag(fasta_oligo[[i]], begin = 180, end = 209)
    myseq <- getSequence(myfrag) %>% paste(collapse = "") %>% toupper()
    fragments <- c(fragments, myseq)
  }
}

# save fragments in a dataframe
df <- data.frame(names, fragments)

# print it
df %>%
  kbl() %>%
  kable_paper("hover", full_width = F)
```

# Save the fragment tables

```{r warning=FALSE}
write.table(df, "../../../output/2_Oligotyping/2E/2E_MATCH_fragments.tsv", sep="\t", quote = FALSE, row.names = FALSE)
write.table(df[,2], "../../../output/2_Oligotyping/2E/2E_only_fragments.tsv", sep="\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

# Do the screening

You need to download the data and process to a quality filtering as described in the manuscript. You also need to create a sample file with the 3 columns : 
- name : sample name of each sample
- r1 : path to R1 file of each sample (after quality filtering)
- r2 : path to R2 file of each sample (after quality filtering)

In the following commands, this file is named 'samples_wolbachia_screening.txt".

```{r engine='bash'}
# load conda environment including illumina-utils
source ~/opt/miniconda3/etc/profile.d/conda.sh
conda activate anvio-7.1

# set paths and variables
FILE="../../../output/2_Oligotyping/2E/2E_MATCH_fragments.tsv"
FILE2="../../../output/2_Oligotyping/2E/2E_only_fragments.tsv"
names=( $(cut -f1 $FILE | sed 1d) )
seq=( $(cut -f1 $FILE2 | sed 1d) )
one=( $(head -n 1 $FILE2) )
seqs=($one ${seq[0]} ${seq[1]} ${seq[2]} ${seq[3]} ${seq[4]} ${seq[5]} ${seq[6]})

# search oligotypes in ovary metagenomes from Reveillaud et al 2019 and Bonneau et al 2018
anvi-script-get-primer-matches --samples-txt ../../../additional_data/samples_wolbachia_screening.txt  \
      --primer-sequences $FILE2 \
      --min-remainder-length 10 \
      --output-dir ../../../../output/2_Oligotyping/2E/

# deactivate conda environment  
conda deactivate

# move results into a separated folder for each oligotype
i=0
for name in "${names[@]}"
  do
    echo $name
    echo ${seqs[i]}
    mv ../../../../output/2_Oligotyping/2E/*${seqs[i]}* ../../../../output/2_Oligotyping/2E/$name
    i=$i+1
  done
```
# Session info
```{r warning=FALSE, message=FALSE}
sessionInfo()
```




