---
title: "Erwinia (oligotyping)"
author: "Hans Schrieke and Julie Reveillaud"
date: "`r Sys.Date()`"
output: 
  rmdformats::readthedown:
    lightbox: true
    gallery: false
    fig_caption: true
    highlight: tango
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load packages, paths, functions
```{r message=FALSE, warning=FALSE}
# Load main packages, paths and custom functions
source("../../../source/main_packages.R")
source("../../../source/functions.R")

# Load supplementary packages
packages <- c("RColorBrewer", "ggpubr", "cowplot", "Biostrings", "openxlsx", "kableExtra")
invisible(lapply(packages, require, character.only = TRUE))
```

# Preparation

## Tables preparation

### Seqtab
```{r warning=FALSE}
# move to oligotyping directory
path_erwinia <- "../../../../output/2_Oligotyping/2A/Erwinia/2A_oligotyping_Erwinia_sequences-c8-s1-a0.0-A0-M10"

# load the matrix count table
matrix_count <- read.table(paste0(path_erwinia, "/MATRIX-COUNT.txt"), header = TRUE) %>% t()

# arrange it
colnames(matrix_count) <- matrix_count[1,]
matrix_count <- matrix_count[-1,]
matrix_count <- matrix_count %>% as.data.frame()

# print it
matrix_count %>%
  kbl() %>%
  kable_paper("hover", full_width = F)
```

### Taxonomy
```{r warning=FALSE}
# load the fasta table
fasta <- readDNAStringSet(paste0(path_erwinia, "/OLIGO-REPRESENTATIVES.fasta"))

# arrange it
fasta <- fasta %>% as.data.frame()
colnames(fasta) <- "seq"
fasta$oligotype <- rownames(fasta)
fasta <- fasta %>% dplyr::select(-c(seq))

# print it
fasta %>%
  kbl() %>%
  kable_paper("hover", full_width = F)
```

### Change oligotype name by oligotype / MED nodes in the matrix count

```{r warning=FALSE}
# Reference file 

## load the reference table
ref_oligo_med2 <- read.table("../../../../output/2_Oligotyping/2B/2B_REF_info_erwinia.tsv", sep="\t", header = TRUE)

## select only the 3 oligotypes of Erwinia
ref_oligo_med2 <- ref_oligo_med2[!is.na(ref_oligo_med2$oligotype),]

## change order of columns
ref_oligo_med2 <- ref_oligo_med2 %>% select(c(seq, oligotype, MED_node_frequency_size, OLIGO_oligotype_frequency_size))

## create a column with reference name (will be used in plots)
ref_oligo_med2$ref <- paste0("oligotype_", ref_oligo_med2$OLIGO_oligotype_frequency_size, " / node_", ref_oligo_med2$MED_node_frequency_size)

## create a copy of fasta 
fasta2 <- fasta

# Matrix count

## create an oligotype column in the matrix count
matrix_count$oligotype <- rownames(matrix_count)

## change order of columns
matrix_count <- matrix_count %>% dplyr::select(c(oligotype, everything()))

## merge the matrix count and the reference dataframe
matrix_count2 <- matrix_count %>% merge(ref_oligo_med2 %>% dplyr::select(-c(seq)), by="oligotype")

## change order of columns
matrix_count2 <- matrix_count2 %>% dplyr::select(c(oligotype, MED_node_frequency_size, OLIGO_oligotype_frequency_size, ref, everything()))

## change rownames
rownames(matrix_count2) <- matrix_count2$ref

## change order of columns
matrix_count2 <- matrix_count2 %>% dplyr::select(-c(oligotype, ref, MED_node_frequency_size, OLIGO_oligotype_frequency_size))

## print it
matrix_count2 %>%
  kbl() %>%
  kable_paper("hover", full_width = F)

## edit the fasta dataframe
fasta2 <- fasta2 %>% merge(ref_oligo_med2 %>% dplyr::select(-c(seq)), by="oligotype")
rownames(fasta2) <- fasta2$ref
fasta2 <- fasta2 %>% dplyr::select(-c(MED_node_frequency_size, OLIGO_oligotype_frequency_size, oligotype))

## print it
fasta2 %>%
  kbl() %>%
  kable_paper("hover", full_width = F)
```
### Metadata
```{r}
metadata <- read.csv("../../../../metadata/metadata.csv", sep=";")
rownames(metadata) <- metadata$Sample
```

### Phyloseq object with oligotypes
```{r warning=FALSE}
# convert matrix_count into matrix and numeric
matrix_count <- matrix_count2 %>% as.matrix()
class(matrix_count) <- "numeric"

# phyloseq elements
OTU = otu_table(as.matrix(matrix_count), taxa_are_rows =TRUE)
TAX = tax_table(as.matrix(fasta2))
SAM = sample_data(metadata)

# phyloseq object
ps <- phyloseq(OTU, TAX, SAM)
ps
compute_read_counts(ps)

# remove blanks
ps <- subset_samples(ps, Strain!="Blank")
ps <- check_ps(ps)
ps

```

### Create new metadata with Percent

#### Load ps with all samples (for final plot)
```{r warning=FALSE}
ps.filter <- readRDS("../../../../output/1_MED/1D/1D_MED_phyloseq_decontam.rds")
ps.filter <- check_ps(ps.filter)
```

#### Edit new metadata with Percent_erwinia
```{r}
guide_italics <- guides(fill = guide_legend(label.theme = element_text(size = 16, face = "italic", colour = "Black", angle = 0)))

# add read depth in sample table of phyloseq object
sample_data(ps.filter)$Read_depth <- sample_sums(ps.filter)

# select Erwinia
ps.erwinia <- ps.filter %>% subset_taxa(Genus=="Erwinia")

# add read depth of Erwinia
sample_data(ps.filter)$Read_erwinia <- sample_sums(ps.erwinia)
sample_data(ps.filter) %>% colnames()
sample_data(ps.erwinia) %>% colnames()

# add percent of Erwinia
sample_data(ps.filter)$Percent_erwinia <- sample_data(ps.filter)$Read_erwinia / sample_data(ps.filter)$Read_depth

# round the percent of Erwinia at 2 decimals
sample_data(ps.filter)$Percent_erwinia <- sample_data(ps.filter)$Percent_erwinia %>% round(2)

# extract metadata table
test <- data.frame(sample_data(ps.filter))

# merge this metadata table with the other
new.metadata <- data.frame(sample_data(ps)) %>% merge(test %>% dplyr::select(c(Sample, Read_depth, Read_erwinia, Percent_erwinia)), by="Sample")
new.metadata <- test[new.metadata$Sample %in% sample_names(ps),]
rownames(new.metadata) <- new.metadata$Sample

# print it
new.metadata %>%
  kbl() %>%
  kable_paper("hover", full_width = F)

# replace metadata in the created phyloseq object
sample_data(ps) <- sample_data(new.metadata)
```

#### Percent by oligotype
```{r}
# Oligotype AAGACTTA

oligo_AAGACTTA <- ps %>% 
  subset_taxa(ref=="oligotype_AAGACTTA (40) | size:38025 / node_N0798 (40) | size:38213")

oligo_AAGACTTA <- prune_taxa(taxa_sums(oligo_AAGACTTA) >= 1, oligo_AAGACTTA)
oligo_AAGACTTA <- prune_samples(sample_sums(oligo_AAGACTTA) >= 1, oligo_AAGACTTA)
  
oligo_AAGACTTA %>% taxa_sums() -> sum_oligo_1
oligo_AAGACTTA@sam_data$Read_depth %>% sum() -> sum_total_1

sum_oligo_1 / sum_total_1 *100 # 4.92%
sum_oligo_1 / 6452623 *100 # 0.59%

data.frame("Sample"=oligo_AAGACTTA@sam_data$Sample, 
           "Read_tot" = oligo_AAGACTTA@sam_data$Read_depth,
           "Read_AAGACTTA"=oligo_AAGACTTA %>% sample_sums(),
           "Percent_AAGACTTA"=round((oligo_AAGACTTA %>% sample_sums() / oligo_AAGACTTA@sam_data$Read_depth *100),2)) -> df1

mean(round((oligo_AAGACTTA %>% sample_sums() / oligo_AAGACTTA@sam_data$Read_depth *100),2))
median(round((oligo_AAGACTTA %>% sample_sums() / oligo_AAGACTTA@sam_data$Read_depth *100),2))

# Oligotype AAGACTTG

oligo_AAGACTTG <- ps %>% 
  subset_taxa(ref=="oligotype_AAGACTTG (22) | size:4594 / node_N0802 (22) | size:4627")

oligo_AAGACTTG <- prune_taxa(taxa_sums(oligo_AAGACTTG) >= 1, oligo_AAGACTTG)
oligo_AAGACTTG <- prune_samples(sample_sums(oligo_AAGACTTG) >= 1, oligo_AAGACTTG)
  
oligo_AAGACTTG %>% taxa_sums() -> sum_oligo_2
oligo_AAGACTTG@sam_data$Read_depth %>% sum() -> sum_total_2

sum_oligo_2 / sum_total_2 *100 # 1.29%
sum_oligo_2 / 6452623 *100 # 0.07%

data.frame("Sample"=oligo_AAGACTTG@sam_data$Sample, 
           "Read_tot" = oligo_AAGACTTG@sam_data$Read_depth,
           "Read_AAGACTTG"=oligo_AAGACTTG %>% sample_sums(),
           "Percent_AAGACTTG"=round((oligo_AAGACTTG %>% sample_sums() / oligo_AAGACTTG@sam_data$Read_depth *100),2)) -> df2

mean(round((oligo_AAGACTTG %>% sample_sums() / oligo_AAGACTTG@sam_data$Read_depth *100),2))
median(round((oligo_AAGACTTG %>% sample_sums() / oligo_AAGACTTG@sam_data$Read_depth *100),2))

# Oligotype TGAGTCGA

oligo_TGAGTCGA <- ps %>% 
  subset_samples(Organ=="Whole" & Strain!="Guadeloupe") %>%
  subset_taxa(ref=="oligotype_TGAGTCGA (29) | size:7110 / node_N0311 (30) | size:8523")

oligo_TGAGTCGA <- prune_taxa(taxa_sums(oligo_TGAGTCGA) >= 1, oligo_TGAGTCGA)
oligo_TGAGTCGA <- prune_samples(sample_sums(oligo_TGAGTCGA) >= 1, oligo_TGAGTCGA)
  
oligo_TGAGTCGA %>% taxa_sums() -> sum_oligo_3 
oligo_TGAGTCGA@sam_data$Read_depth %>% sum() -> sum_total_3 

sum_oligo_3 / sum_total_3 *100 # 1.36%
sum_oligo_3 / 6452623 *100 # 0.11%

data.frame("Sample"=oligo_TGAGTCGA@sam_data$Sample, 
           "Read_tot" = oligo_TGAGTCGA@sam_data$Read_depth,
           "Read_TGAGTCGA"=oligo_TGAGTCGA %>% sample_sums(),
           "Percent_TGAGTCGA"=round((oligo_TGAGTCGA %>% sample_sums() / oligo_TGAGTCGA@sam_data$Read_depth *100),2)) -> df3

mean(round((oligo_TGAGTCGA %>% sample_sums() / oligo_TGAGTCGA@sam_data$Read_depth *100),2))
median(round((oligo_TGAGTCGA %>% sample_sums() / oligo_TGAGTCGA@sam_data$Read_depth *100),2))
```


# Taxonomic structure

## Count

```{r fig.height=12, fig.width=14}

col <- brewer.pal(7, "Pastel2")

# reshape data for plot
test3 <- test %>% select(c(Sample, Species, Strain, Organ, Read_depth, Read_erwinia)) %>% reshape2::melt(id.vars=c("Sample", "Species", "Strain", "Organ"), vars=c("Read_depth", "Read_erwinia"))

count_whole <- test3[test3$Organ=="Whole",]
count_ovary <- test3[test3$Organ=="Ovary",]

make.italic <- function(x) as.expression(lapply(x, function(y) bquote(italic(.(y)))))

levels(count_whole$Species)= c("Aedes aegypti"=make.italic("Aedes aegypti"),
               "Culex pipiens"=make.italic("Culex pipiens"),
               "Culex quinquefasciatus"=make.italic("Culex quinquefasciatus"))

levels(count_ovary$Species)= c("Aedes aegypti"=make.italic("Aedes aegypti"),
               "Culex pipiens"=make.italic("Culex pipiens"),
               "Culex quinquefasciatus"=make.italic("Culex quinquefasciatus"))

levels(count_whole$Strain) <- c("Bosc", "Camping~Europe", "Guadeloupe", "Lavar~(lab)", expression(paste(italic("Wolbachia"), "- (Slab TC)")))

levels(count_ovary$Strain) <- c("Bosc", "Camping~Europe", "Guadeloupe", "Lavar~(lab)", expression(paste(italic("Wolbachia"), "- (Slab TC)")))


# plot
p_count1 <- ggplot(count_whole, aes(x = Sample, y = value, fill=variable))+ 
  geom_bar(position = "dodge", stat = "identity")+
  scale_fill_manual(values = col)+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size=12, hjust=1, vjust=0.5)) +
  ggtitle("") + 
  guide_italics+
  theme(legend.title = element_text(size = 20), 
        legend.position="bottom",
        legend.text=element_text(size=14), 
        panel.spacing.y=unit(1, "lines"), 
        panel.spacing.x=unit(0.8, "lines"),
        panel.spacing=unit(0,"lines"),
        strip.background=element_rect(color="grey30", fill="grey90"),
        strip.text.x = element_text(size = 16),
        panel.border=element_rect(color="grey90"),
        axis.ticks.x=element_blank(),
        axis.text.y = element_text(size=18)) +
  facet_wrap(~Species+Strain+Organ, scales = "free_x", ncol=3, labeller=label_parsed)+
  labs(y="Sequence counts")+
  ylim(0, 900000)+
  geom_text(aes(label=value), position=position_dodge(width=1.1), width=0.25, size=4, hjust=-0.25, vjust=0.5, angle=90)+
  guides(fill=guide_legend(title="Read"))

p_count2 <- ggplot(count_ovary, aes(x = Sample, y = value, fill=variable))+ 
  geom_bar(position = "dodge", stat = "identity")+
    scale_fill_manual(values = col)+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size=18, hjust=1, vjust=0.5)) +
  ggtitle("") + 
  guide_italics+
  theme(legend.title = element_text(size = 20), 
        legend.position="bottom",
        legend.text=element_text(size=14), 
        panel.spacing.y=unit(1, "lines"), 
        panel.spacing.x=unit(0.8, "lines"),
        panel.spacing=unit(0,"lines"),
        strip.background=element_rect(color="grey30", fill="grey90"),
        strip.text.x = element_text(size = 16),
        panel.border=element_rect(color="grey90"),
        axis.ticks.x=element_blank(),
        axis.text.y = element_text(size=18)) +
 facet_wrap(~Species+Strain+Organ, scales = "free_x", ncol=3, labeller=label_parsed)+
  labs(y="Sequence counts")+
    ylim(0, 900000)+
  geom_text(aes(label=value), position=position_dodge(width=0.8), width=0.25, size=4, hjust=-0.25, vjust=0.5, angle=90)+
  guides(fill=guide_legend(title="Read"))

# afficher plot
p_count1
p_count2

# panels
p_group <- plot_grid(p_count1+theme(legend.position="none"), 
          p_count2+theme(legend.position="none"), 
          nrow=2, 
          ncol=1)+
    draw_plot_label(c("B1", "B2"), c(0, 0), c(1, 0.5), size = 20)

legend_plot <- get_legend(p_count1 + theme(legend.position="bottom"))

p_counts <- plot_grid(p_group, legend_plot, nrow=2, ncol=1, rel_heights = c(1, .1))
p_counts
```

## Whole (the most abundant nodes)

```{r fig.height = 12, fig.width=18}
guide_italics <- guides(fill = guide_legend(label.theme = element_text(size = 16, face = "italic", colour = "Black", angle = 0),
                                            nrow=2, byrow=TRUE))

# select whole
ps.filter.whole <- subset_samples(ps, Organ=="Whole")
ps.filter.whole <- prune_taxa(taxa_sums(ps.filter.whole) >= 1, ps.filter.whole)
ps.filter.whole <- prune_samples(sample_sums(ps.filter.whole) >= 1, ps.filter.whole)
ps.filter.whole

# data pour plot
#data_for_plot2 <- taxo_data_fast(ps.filter.whole, method = "abundance")
data_for_plot2 <- taxo_data(ps.filter.whole, top=15)
paste0("\n15 MOST ABUNDANT GENUS: \n") %>% cat()
paste0("\"", levels(data_for_plot2$Name), "\",\n") %>% cat()

data_for_plot2$Name <- data_for_plot2$Name %>% gsub(pattern = "node_", replacement ="" ) %>% as.factor()
data_for_plot2$Name <- as.factor(data_for_plot2$Name)

new_names <- c("oligotype_AAGACTTA (40) | size:38025 / N0798 (40) | size:38213.",
               "oligotype_TGAGTCGA (29) | size:7110 / N0311 (30) | size:8523.",
               "oligotype_AAGACTTG (22) | size:4594 / N0802 (22) | size:4627.",
               "Other.")

data_for_plot2$Name <- factor(data_for_plot2$Name, levels = new_names)

col_add <- brewer.pal(8, "Accent")

col <- c("oligotype_AAGACTTA (40) | size:38025 / N0798 (40) | size:38213."="#FFFFCF",
         "oligotype_TGAGTCGA (29) | size:7110 / N0311 (30) | size:8523."="#FFE352",
         "oligotype_AAGACTTG (22) | size:4594 / N0802 (22) | size:4627."="#F5F61B",
         "Other."="#A0A0A0")

levels(data_for_plot2$Species)= c("Aedes aegypti"=make.italic("Aedes aegypti"),
               "Culex pipiens"=make.italic("Culex pipiens"),
               "Culex quinquefasciatus"=make.italic("Culex quinquefasciatus"))

levels(data_for_plot2$Strain) <- c("Bosc", "Camping~Europe", "Guadeloupe", "Lavar~(lab)", expression(paste(italic("Wolbachia"), "- (Slab TC)")))

#data_for_plot2 <- data_for_plot2 %>% na.omit()

p2 <- ggplot(data_for_plot2, aes(x = Sample, y = Relative_Abundance, fill = Name, species=Species, organ=Organ, Strain=Strain))+ 
  geom_bar(position = "stack", stat = "identity")+
  scale_fill_manual(values = col)+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size=18, hjust=1, vjust=0.5)) +
  ggtitle("") + 
  guide_italics+
  theme(legend.title = element_text(size = 20), 
        legend.position="bottom",
        legend.text = element_text(size=14),
        #legend.key.height = unit(1, 'cm'),
        panel.spacing.y=unit(1, "lines"), 
        panel.spacing.x=unit(0.8, "lines"),
        panel.spacing=unit(0,"lines"),
        strip.background=element_rect(color="grey30", fill="grey90"),
        strip.text.x = element_text(size = 16),
        panel.border=element_rect(color="grey90"),
        axis.ticks.x=element_blank(),
        axis.text.y = element_text(size=18)) +
  facet_wrap(~Species+Strain+Organ, scales = "free", ncol=3, labeller=label_parsed)+
  labs(x="Sample", y="Relative abundance", fill="Oligotype / MED node")

p2
```




## Ovary (the most abundant nodes)

```{r fig.height = 12, fig.width=18}
# select ovary
ps.filter.ovary <- subset_samples(ps, Organ=="Ovary")
ps.filter.ovary <- prune_taxa(taxa_sums(ps.filter.ovary) >= 1, ps.filter.ovary)
ps.filter.ovary <- prune_samples(sample_sums(ps.filter.ovary) >= 1, ps.filter.ovary)
ps.filter.ovary

# data pour plot
data_for_plot3 <- taxo_data(ps.filter.ovary, top=15)
paste0("\n15 MOST ABUNDANT GENUS: \n") %>% cat()
paste0("\"", levels(data_for_plot3$Name), "\",\n") %>% cat()

data_for_plot3$Name <- data_for_plot3$Name %>% gsub(pattern = "node_", replacement ="" ) %>% as.factor()
data_for_plot3$Name <- as.factor(data_for_plot3$Name)

levels(data_for_plot3$Species)= c("Culex pipiens"=make.italic("Culex pipiens"))

levels(data_for_plot3$Strain) <- c("Bosc", "Camping~Europe",  "Lavar~(lab)")

p3 <- ggplot(data_for_plot3, aes(x = Sample, y = Relative_Abundance, fill = Name, species=Species, organ=Organ, Strain=Strain))+ 
  geom_bar(position = "stack", stat = "identity")+
  scale_fill_manual(values = col)+
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size=18, hjust=1, vjust=0.5)) +
  ggtitle("") + 
  guide_italics+
  theme(legend.title = element_text(size = 20), 
        legend.position="bottom",
        legend.text = element_text(size=14),
        #legend.key.height = unit(1, 'cm'),
        panel.spacing.y=unit(1, "lines"), 
        panel.spacing.x=unit(0.8, "lines"),
        panel.spacing=unit(0,"lines"),
        strip.background=element_rect(color="grey30", fill="grey90"),
        strip.text.x = element_text(size = 16),
        panel.border=element_rect(color="grey90"),
        axis.ticks.x=element_blank(),
        axis.text.y = element_text(size=18)) +
  facet_wrap(~Species+Strain+Organ, scales = "free", ncol=3, labeller=label_parsed)+
  labs(x="Sample", y="Relative abundance", fill="Oligotype / MED node")

p3
```

## Panels taxonomy of whole / ovary
```{r}
legend_plot <- get_legend(p2 + theme(legend.position="bottom"))

# panels
p_group <- plot_grid(p2+theme(legend.position="none"), 
          p3+theme(legend.position="none"), 
          nrow=2, 
          ncol=1)+
    draw_plot_label(c("A1", "A2"), c(0, 0), c(1, 0.5), size = 20)

p_taxo <- plot_grid(p_group, legend_plot, nrow=2, rel_heights = c(1, .1))
p_taxo
```

## Save taxonomic plot

```{r warning=FALSE}
tiff("../../../../output/2_Oligotyping/2D/2De_OLIGO_counts_erwinia.tiff", units="in", width=20, height=18, res=300)
p_counts
dev.off()

tiff("../../../../output/2_Oligotyping/2D/2De_OLIGO_taxonomic_erwinia_whole.tiff", units="in", width=16, height=12, res=300)
p2
dev.off()

tiff("../../../../output/2_Oligotyping/2D/2De_OLIGO_taxonomic_erwinia_ovary.tiff", units="in", width=18, height=14, res=300)
p3
dev.off()

tiff("../../../../output/2_Oligotyping/2D/2De_OLIGO_taxonomic_erwinia.tiff", units="in", width=18, height=16, res=300)
p_taxo
dev.off()

png("../../../../output/2_Oligotyping/2D/2De_OLIGO_counts_erwinia_big.png", units="in", width=20, height=18, res=300)
p_counts
dev.off()

png("../../../../output/2_Oligotyping/2D/2De_OLIGO_counts_erwinia_small.png", units="in", width=18, height=14, res=300)
p_counts
dev.off()

png("../../../../output/2_Oligotyping/2D/2De_OLIGO_taxonomic_erwinia_whole.png", units="in", width=16, height=12, res=300)
p2
dev.off()

png("../../../../output/2_Oligotyping/2D/2De_OLIGO_taxonomic_erwinia_ovary.png", units="in", width=18, height=14, res=300)
p3
dev.off()

png("../../../../output/2_Oligotyping/2D/2De_OLIGO_taxonomic_erwinia_big.png", units="in", width=18, height=18, res=300)
p_taxo
dev.off()

png("../../../../output/2_Oligotyping/2D/2De_OLIGO_taxonomic_erwinia_small.png", units="in", width=18, height=14, res=300)
p_taxo
dev.off()
```

# Make main plot
```{r warning=FALSE, fig.height = 12, fig.width=18}
img <- magick::image_read(paste0(path_erwinia, "/HTML-OUTPUT/entropy.png"))
p_entropy <- magick::image_ggplot(img, interpolate = TRUE)
p_entropy+ theme(plot.margin = unit(c(-7,-2.5,-7,-0.5), "cm"))
p_entropy+ theme(plot.margin=unit(c(-7,-2,-12,-5), "mm"))

aligned <- plot_grid(p_taxo, 
                     p_counts, 
                     align="hv")

aligned

p_entropy2 <- plot_grid(p_entropy, nrow=1)+
  draw_plot_label(c("C"), c(0), c(1), size=20, hjust=-0.5)

p_entropy2

t_plot <- plot_grid(aligned, 
                    p_entropy2,
                    nrow=2, 
                    ncol=1, 
                    scale=1,
                    rel_heights=c(2,1))

t_plot

tiff("../../../../output/2_Oligotyping/2D/2De_OLIGO_main_erwinia.tiff", width=36, height=36, res=300, units="in")
t_plot
dev.off()

png("../../../../output/2_Oligotyping/2D/2De_OLIGO_main_erwinia.png", width=36, height=36, res=300, units="in")
t_plot
dev.off()
```


# Session info
```{r warning=FALSE, message=FALSE}
sessionInfo()
```





